packages/components/src/internal/components/chart/ChartBuilderModal.tsx

```diff
@@ -0,0 +1,648 @@
+import React, { FC, Fragment, memo, useCallback, useEffect, useMemo, useRef, useState } from 'react';
+import classNames from 'classnames';
+
+import { PermissionTypes, Utils } from '@labkey/api';
+
+import { generateId } from '../../util/utils';
+import { LABKEY_VIS } from '../../constants';
+import { Modal } from '../../Modal';
+
+import { SelectInput, SelectInputOption } from '../forms/input/SelectInput';
+
+import { naturalSortByProperty } from '../../../public/sort';
+
+import { LoadingSpinner } from '../base/LoadingSpinner';
+
+import { deleteChart, saveChart } from '../../actions';
+
+import { QueryModel } from '../../../public/QueryModel/QueryModel';
+import { RequiresModelAndActions } from '../../../public/QueryModel/withQueryModels';
+
+import { useServerContext } from '../base/ServerContext';
+import { hasPermissions } from '../base/models/User';
+
+import { Alert } from '../base/Alert';
+import { FormButtons } from '../../FormButtons';
+
+import { getContainerFilter } from '../../query/api';
+
+import { SVGIcon } from '../base/SVGIcon';
+
+import { ChartConfig, ChartQueryConfig, GenericChartModel } from './models';
+
+interface AggregateFieldInfo {
+    name: string;
+    value: string;
+}
+
+export interface ChartFieldInfo {
+    aggregate?: AggregateFieldInfo;
+    // allowMultiple?: boolean;
+    label: string;
+    name: string;
+    nonNumericOnly?: boolean;
+    numericOnly?: boolean;
+    required: boolean;
+}
+
+export interface ChartTypeInfo {
+    fields: ChartFieldInfo[];
+    hidden?: boolean;
+    imgUrl: string;
+    name: string;
+    title: string;
+}
+
+const HIDDEN_CHART_TYPES = ['time_chart'];
+const RIGHT_COL_FIELDS = ['color', 'shape', 'series'];
+export const MAX_ROWS_PREVIEW = 100000;
+export const MAX_POINT_DISPLAY = 10000;
+const BLUE_HEX_COLOR = '3366FF';
+
+const ICONS = {
+    bar_chart: 'bar_chart',
+    box_plot: 'box_plot',
+    pie_chart: 'pie_chart',
+    scatter_plot: 'xy_scatter',
+    line_plot: 'xy_line',
+};
+
+interface Props extends RequiresModelAndActions {
+    onHide: (successMsg?: string) => void;
+    savedChartModel?: GenericChartModel;
+}
+
+export const ChartBuilderModal: FC<Props> = memo(({ actions, model, onHide, savedChartModel }) => {
```

Comment:
My general feedback for this component is that the body of it is pretty complex, it's doing a lot (because it's designed to do a lot) and I think would benefit from being split into a few smaller components that can encapsulate the logic for each part. Especially considering that we're going to be adding to this thing over time, and it will only get more complicated. The "easy" divisions I see would be to split it into:

1. SideBar
2. ConfigForm
3. Preview
4. Footer

But there could definitely be other ways to split this up. I don't think we need a new file for each component.

Reply:
done